Prompt when the destination readme has been edited and provide `--yes` flag for automation

infiniteluke opened the pull request:
<!--
Thanks for your interest in the project. We appreciate bugs filed and PRs submitted!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
**What**:
Add a warning if someone edited the README manually. Currently emdaer overwrites the file without warning. Also, new users often accidentally edit the README directly before realizing emdaer is installed.

I have it implemented as a warning that does not fail the process.
<img width="913" alt="screen shot 2018-03-05 at 10 05 07 pm" src="https://user-images.githubusercontent.com/1127238/37016490-5175f2d4-20c1-11e8-922d-8165080779af.png">

It could be argued we should error here instead.


<!-- Why are these changes necessary? -->
**Why**:
better ux

<!-- How were these changes implemented? -->
**How**:
execute git to check if the file being generated already exists and has a change.

<!-- Have you done all of these things?  -->
**Checklist**:
<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->
- [x] Documentation
- [x] Tests
- [x] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->
- [x] Added myself to AUTHORS <!-- this is optional, see the contributing guidelines for instructions -->
- [x] Change `--path` to emdaer taking space separated paths as per lint staged https://github.com/okonet/lint-staged/issues/138#issue-213736444.
- [x] Document path usage and lint-staged usage
<!-- Feel free to add additional comments -->

infiniteluke committed:
feat: warn when the destination readme has been edited

flipactual commented:
Yeah, we should error – emdaer is not able to do its thing, so, yeah, full on error

I'm not clear on what happens in the scenario where someone modifies `.emdaer/README.emdaer.md`, runs emdaer, then adds their changes and tries to commit – does this work assume that emdaer is only being run as a commit hook?

infiniteluke commented:
> runs emdaer, then adds their changes and tries to commit

That's a good point @flipactual.. We may want to put this behind a flag. So the default would be that it does not check for changes. But, if you want emdaer to be `strict` you can enable strict mode with a flag? I was trying to avoid adding a flag because I like the opinionated nature of emdaer currently.

Still though, it seems like this feature could make incremental runs problematic. Is there a simple way to solve for this?

flipactual commented:
The simplest method I can think of at the moment is to log with the error instructions like "hey, you have manual changes in this file, do a git checkout for that file and try again"

infiniteluke commented:
Ok - I'm going to try something to that affect. But I also want to mention something about "moving changes to {NAME}.emdaer.md if you edited it manually"

flipactual commented:
Yeah, that's good, just draw attention to the fact that you need to checkout the file before emdaer will be happy

infiniteluke commented:
Another way to handle this would be with a prompt.

```
$ emdaer --checkDirtyDestination
  Are you sure you want to overwrite {NAME}.md. It has existing changes. y\n
$ y
  emdaer did the thing
```

a `--yes` flag could be used to force the overwrite for precommit hooks or other automation.

We could work out whether having the default be overwriting or prompting, but that'd be a decent solution.

Is that worth building before we the `plz git checkout that file` solution?

flipactual commented:
Either way, I think this counts as a breaking change... if you want to go all the way with it, go for it

infiniteluke commented:
If by default emdaer still overwrites, I don't think this is breaking. 

Should emdaer default to prompting or overwritting? I think UX tells me it should prompt by default since compiling readmes is new to most people. Maybe we can ask around a bit.

flipactual commented:
I think it should default to prompting

And I think we should do a proper check to see if generated files have changed. We'll need to compare hashes instead of just checking git

infiniteluke commented:
If the generated file hasn't changed should we simply not write the "new" one and skip logging?

flipactual commented:
Sounds fine

infiniteluke committed:
feat(cli): add prompt and --yes flag for overwriting readmes

BREAKING CHANGES: defaults to prompt Y/n if a readme has existing changes

infiniteluke commented:
<img width="1280" alt="emdaer" src="https://user-images.githubusercontent.com/1127238/37074083-b76ea34c-217f-11e8-931d-350e8384b6c8.png">

infiniteluke commented:
Open to wordsmithing plz

infiniteluke commented:
I need to document this in the readme before we merge

infiniteluke committed:
docs: add note about prompt and --yes flag

infiniteluke commented:
We shouldn't use git for tracking changes. We will store an md5 hash of the readme contents in a namespaced comment at the top of the file. Before emdaer runs, we will hash the content and compare it to the hash we found in the file. if there is a difference, we should prompt for overwriting.

After emdaer runs we will hash the generated contents, if that hash matches the prior hash, no changes occurred and we can skip writing the file.

infiniteluke commented:
As discussed on zoom:

- partitioned prompts should be executed serially each awaiting the prior, while unprompted files are processed behind the scenes.
- logs should not be printed until prompts have been answered
- the above code is half refactored. ex: the `exitCode` isn't even returned at this point.
- hash implementation works but is currently in a broken state due to the stream wrangling I couldn't figure out.

infiniteluke committed:
feat: simplify cli

closes #89
closes #96

BREAKING CHANGE: @emdaer/cli now prompts the user if a manual change to the compiled README is detected. Use --yes to skip the prompt and automatically overwrite the README

infiniteluke commented:
> partitioned prompts should be executed serially each awaiting the prior, while unprompted files are processed behind the scenes.

I did not complete this piece, but still think we can ship as is. I'm happy to open a ticket for this feature, as it will be a non breaking change.

I opened https://github.com/emdaer/emdaer/issues/97 for this

infiniteluke commented:
# BREAKING CHANGE
@emdaer/cli now prompts the user if a manual change to the compiled README is detected. Use --yes to skip the prompt and automatically overwrite the README

Merging this PR should cut 3.0.0. But who knows 😂

infiniteluke committed:
build: remove unused flow-typed defs

flipactual commented on packages/cli/src/cli.js:
```diff
@@ -0,0 +1,105 @@
+const glob = promisify(require('glob'));
+const { outputFile, readFile, readJson } = require('fs-extra');
+const inquirer = require('inquirer');
+const Promise = require('bluebird');
```
y tho?

infiniteluke replied:
`Promise.each` makes the sequential prompt feature super easy

flipactual replied:
Need this to happen... https://github.com/tc39/proposal-async-iteration

infiniteluke commented:
_moved to main PR comment_

infiniteluke committed:
feat: allow emdaer to take path parameters and document lint-staged usage

infiniteluke committed:
docs: document manual usage of emdaer
